### PR TITLE
Create a minimal library

### PR DESCRIPTION
**Pull request by procrastinationfighter**
Fixes: #12 
This PR transforms the printer's code into the library's interface. 
In order to achieve this, the `StreamReader` was modified and now accepts a `Consumer` instance instead of a callback function.
The Printer was refactored and now uses the created interface.

### CODE REVIEW

File: scylla-cdc/src/cdc_reader.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+use crate::consumer::ConsumerFactory;
+use crate::reader::StreamReader;
+use crate::stream_generations::GenerationFetcher;
+
+pub struct CDCReader {
+    // Tells the worker to stop
+    // Usage of the "watch" channel will make it possible to change the the timestamp later,
+    // for example if somebody loses patience and wants to stop now not later
+    end_timestamp: tokio::sync::watch::Sender<Duration>,
+}
```
Comment (procrastinationfighter) [inline scylla-cdc/src/cdc_reader.rs:22]: I don't like the name `Reader`, because it may suggest that it's something similar to `StreamReader`, but I couldn't come up with a better name - I'm open for suggestions.

File: scylla-cdc/src/cdc_reader.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            start_timestamp,
+            end_timestamp,
+            window_size,
+            safety_interval,
+            sleep_interval,
+            end_timestamp_receiver,
+            consumer_factory,
+        );
+        let (_, handle) = tokio::task::spawn(async move { worker.run().await }).remote_handle();
+        let printer = CDCReader {
```
Comment (procrastinationfighter) [inline scylla-cdc/src/cdc_reader.rs:53]: This piece of code (spawning) required the `ConsumerFactory` object to implement also a `Send` trait. From what I've seen, it's probably automatically generated if all members of a structure implement Send, so it's probably not a bothering issue, is it?

File: scylla-cdc/src/cdc_reader.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                            let keyspace = self.keyspace.clone();
+                            let table_name = self.table_name.clone();
+                            let consumer = self.consumer_factory.new_consumer();
+                            tokio::spawn(async move {
+                                reader
+                                    .fetch_cdc_with_consumer(keyspace, table_name, consumer)
+                                    .await
+                            })
+                        })
+                        .collect();
```
Comment (procrastinationfighter) [inline scylla-cdc/src/cdc_reader.rs:199]: I didn't manage to create a new consumer asynchronously, so I removed the `async` trait from `ConsumerFactory`. An alternative is to use [block_on](https://docs.rs/futures/0.2.0/futures/executor/fn.block_on.html), but I think that in this way we show the user that it doesn't matter if the consumers can be created asynchronously.

Comment (procrastinationfighter) [reply scylla-cdc/src/cdc_reader.rs:199]: Also, the spawning forced me to require consumers to implement `Send`. Just like with `ConsumerFactory` (see my other comment).

File: scylla-cdc/src/reader.rs [hunk 4]
```diff
@@ -101,6 +102,78 @@ impl StreamReader {
...
+        &self,
+        keyspace: String,
+        table_name: String,
+        mut consumer: Box<dyn Consumer + Send>,
+    ) -> anyhow::Result<()> {
+        // [TODO]: This function is almost identical to fetch_cdc,
+        //      but it uses the Consumer trait as a callback.
+        //      The reason for leaving the old function are the tests,
+        //      there is no obvious solution as for what to do with them.
+
```
Comment (procrastinationfighter) [inline scylla-cdc/src/reader.rs:116]: This temporary solution is left, because I am not sure what to do with the old tests - if the old `fetch_cdc` function is removed, we will have to fix them.

File: scylla-cdc-printer/src/printer.rs [hunk 5]
```diff
@@ -1,220 +1,44 @@
...
+impl Consumer for PrinterConsumer {
+    async fn consume_cdc(&mut self, data: CDCRow<'_>) -> anyhow::Result<()> {
+        // [TODO]: Add prettier printing
+        println!("time: {}, batch_seq_no: {}, end_of_batch: {}, operation: {:?}, ttl: {:?}",
+                 data.time,
+                 data.batch_seq_no,
+                 data.end_of_batch,
+                 data.operation,
+                 data.ttl);
+        Ok(())
```
Comment (procrastinationfighter) [inline scylla-cdc-printer/src/printer.rs:17]: I suggest leave the pretty printing for another PR.

File: scylla-cdc/src/cdc_reader.rs (hunk 1 above)
Comment (piodul) [reply scylla-cdc/src/cdc_reader.rs:22]: It reads a whole CDC log, so maybe `CDCLogReader`?

File: scylla-cdc/src/cdc_reader.rs (hunk 2 above)
Comment (piodul) [reply scylla-cdc/src/cdc_reader.rs:53]: Yes, requiring `Send` is fine. However, I think that `ConsumerFactory` should require `Sync`, which is a stronger trait than `Send`. This will allow the factory to be shared by multiple tasks and you will be able to keep `ConsumerFactory::new_consumer` async and make it possible to call it concurrently.

File: scylla-cdc/src/reader.rs (hunk 4 above)
Comment (piodul) [reply scylla-cdc/src/reader.rs:116]: Won't just passing a lambda to `fetch_cdc` which calls the `consumer.consume_cdc` work? Something like that (in `CDCReaderWorker::run`):

```rust
let reader = Arc::clone(reader);
let keyspace = self.keyspace.clone();
let table_name = self.table_name.clone();
let mut consumer = self.consumer_factory.new_consumer();
tokio::spawn(async move {
    reader
        .fetch_cdc(keyspace, table_name, |row| consumer.consume_cdc(row))
        .await
})
```

Comment (piodul) [reply scylla-cdc/src/reader.rs:116]: The `fetch_cdc` will have to be modified so that the lambda/closure is able to return `anyhow::Result<()>`.

File: scylla-cdc/src/cdc_reader.rs (hunk 3 above)
Comment (piodul) [reply scylla-cdc/src/cdc_reader.rs:199]: Consumers can be `Send`, it's practically required when you work with `tokio` and want an object alive across `.await` points because `tokio`'s scheduler is allowed to move tasks between worker threads.

I'd recommend putting the `Sync` and `Send` requirements in the traits' definitions in order to simplify things:

```rust
#[async_trait]
pub trait Consumer: Send {
    /* ... */
}

#[async_trait]
pub trait ConsumerFactory: Sync {
    /* ... */
}
```

**Comment by procrastinationfighter** [comment]
Ok, but we need to transform the row into a `CDCRow`, and to do it we need
the schema, and to create the schema, we need to run at least one query to
get the specs. I'm not sure how to solve this issue without doing a dummy
query in the `run` function.
(I'm sorry if this message appears in some weird place, I'm using the email
interface for the first time)

Piotr Dulikowski ***@***.***> schrieb am Mo., 4. Apr. 2022,
19:52:

**Comment by piodul** [comment]
> Ok, but we need to transform the row into a `CDCRow`, and to do it we need the schema, and to create the schema, we need to run at least one query to get the specs. I'm not sure how to solve this issue without doing a dummy query in the `run` function. (I'm sorry if this message appears in some weird place, I'm using the email interface for the first time)

OK, I see the problem now. Let's start with adapting the existing `fetch_cdc` function and the tests so that the callback takes a `CDCRow` instead of `Row`, then rebase the code from the PR on top of that.

You will probably have to modify the `fetch_callback` quite a bit - you won't be able to push the CDCRow into an outer vector because of lifetimes - so you will probably have to extract values from the CDC row and convert it into a tuple, struct or something else before putting it into the vector.

File: scylla-cdc/src/log_reader.rs [hunk 6]
```diff
@@ -0,0 +1,213 @@
+use std::sync::Arc;
+
+use anyhow;
+use chrono::Duration;
+use core::time;
```
Comment (piodul) [inline scylla-cdc/src/log_reader.rs:5]: Importing from `core` is a bit strange. This should be the same as `use std::time;`.

File: scylla-cdc/src/log_reader.rs [hunk 7]
```diff
@@ -0,0 +1,213 @@
+use std::sync::Arc;
+
+use anyhow;
+use chrono::Duration;
```
Comment (piodul) [inline scylla-cdc/src/log_reader.rs:4]: Let's import the `std::time::Duration` directly and use qualified `chrono::Duration`. I think it is more common for an unqualified `Duration` to come from `std`, not `chrono`, so this should be less confusing.

File: scylla-cdc/src/log_reader.rs [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    safety_interval: Duration,
+    readers: Vec<Arc<StreamReader>>,
+    end_timestamp_receiver: tokio::sync::watch::Receiver<Duration>,
+    consumer_factory: Arc<dyn ConsumerFactory>,
+}
+
+impl CDCReaderWorker {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        session: &Arc<Session>,
```
Comment (piodul) [inline scylla-cdc/src/log_reader.rs:88]: Nitpick: it's better to accept by value if you are going to clone anyway. Now, there is a mandatory clone of the `Arc`, which involves a relatively expensive atomic operations, but if you accept by value then the caller will be able to choose if they want to move the `Arc` in or clone it.

File: scylla-cdc/src/log_reader.rs [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+                    // that could have been missed because of earlier fetching failures.
+                    // More on this here https://github.com/piodul/scylla-cdc-rust/pull/10#discussion_r826865162
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    async fn set_upper_timestamp(readers: &[Arc<StreamReader>], new_upper_timestamp: Duration) {
+        // This function does not take &self, because that would require the CDCWorker to be Send.
```
Comment (piodul) [inline scylla-cdc/src/log_reader.rs:208]: Do you know why `CDCReaderWorker` is not `Send`? The printer worker had a similar function and it worked fine.

File: scylla-cdc-printer/src/printer.rs [hunk 10]
```diff
@@ -1,220 +1,42 @@
...
-    table_name: String,
-    start_timestamp: Duration,
-    end_timestamp: Duration,
-    sleep_interval: time::Duration,
-    window_size: Duration,
-    safety_interval: Duration,
-    readers: Vec<Arc<StreamReader>>,
-    end_timestamp_receiver: tokio::sync::watch::Receiver<Duration>,
-}
+struct PrinterConsumerFactory {}
```
Comment (piodul) [inline scylla-cdc-printer/src/printer.rs:19]: Nit: you can avoid the brackets after the struct name if it has no fields. This is called a "unit" struct. IMO it looks a bit nicer when you don't have to include the brackets every time you instantiate such struct.

File: scylla-cdc-printer/src/printer.rs [hunk 11]
```diff
@@ -1,220 +1,42 @@
...
-            start_timestamp,
-            end_timestamp,
-            window_size,
-            safety_interval,
-            sleep_interval,
-            end_timestamp_receiver,
+use async_trait::async_trait;
+use scylla_cdc::consumer::{CDCRow, Consumer, ConsumerFactory};
+
+struct PrinterConsumer {}
```
Comment (piodul) [inline scylla-cdc-printer/src/printer.rs:5]: Same tip as for `PrinterConsumerFactory`.

File: scylla-cdc-printer/src/printer.rs [hunk 12]
```diff
@@ -1,220 +1,42 @@
...
+impl ConsumerFactory for PrinterConsumerFactory {
+    async fn new_consumer(&self) -> Box<dyn Consumer> {
+        Box::new(PrinterConsumer {})
     }
 }
 
 #[cfg(test)]
 mod tests {
+    use chrono::Duration;
+    use core::time;
```
Comment (piodul) [inline scylla-cdc-printer/src/printer.rs:31]: `use std::time;`

**Review by piodul — COMMENTED** [review]
Left several small nitpicks.

In one commit, you copy the printer and immediately adapt it into a library, and in the next commit you delete most of the code of the printer. Because of that it's hard to verify by looking at the diffs what modifications have been made to the printer. Please try making the modifications more clear, e.g. first refactor the CDCLogPrinter into CDCLogReader but keep both in the same file, then move the CDCLogReader into a separate file.

Apart from that, I think it looks good.

File: scylla-cdc/src/log_reader.rs (hunk 9 above)
Comment (procrastinationfighter) [reply scylla-cdc/src/log_reader.rs:208]: It was probably a relict from the times when `ConsumerFactory` wasn't `Sync` and I forgot to change it, my bad.

**Comment by procrastinationfighter** [comment]
I'm not sure if I modified the history the way you wished, but now when I look into the `Commits` tab on Github, it looks better.
Also, a reminder: these changes have not been tested correctly yet, so some bugs might appear after using the tests from #46.

**Comment by piodul** [comment]
> I'm not sure if I modified the history the way you wished, but now when I look into the `Commits` tab on Github, it looks better. Also, a reminder: these changes have not been tested correctly yet, so some bugs might appear after using the tests from #46.

Yes, that's more or less what I requested, thanks. Let's fix the bugs that will appear along with the end-to-end test pull request. Some bugs may also appear after we run the `scylla-test`.

**Review by piodul — APPROVED** [review]
